### PR TITLE
CNF-22150: Make Uncore multi-node

### PR DESCRIPTION
**Pull request by irinamihai**
Description:
- Add nodeSelector field to UncoreSpec for targeting specific nodes
- Add conflict resolution: active config stays, oldest-wins tiebreaker when multiple Uncores match a node
- Add generic helpers filterByNodeSelector, selectActiveOrOldest, and getActiveResourceName in common.go for node matching and conflict resolution
- Add NodeInfo (CPU capacity, architecture) to PowerNodeState as SSA anchor to prevent null status on last field removal
- Add envtest reconcile tests covering single CR, conflict resolution, deletion, label changes, die/package tuning, empty selector, sticky config, and config updates
- Remove StatusErrors from UncoreStatus — all status is reported via PowerNodeState

Note: not all cases when values are missing from the Uncore CR are handled. This is an existing issue. To keep it consistent with the PowerProfile, the hardware min/max should be used. An issue will be created in the repo to track this.

### CODE REVIEW

File: controllers/common.go [hunk 1]
```diff
@@ -50,75 +49,126 @@ func powerProfileFieldManager(profileName string) string {
...
 		logger.V(5).Info("No label selector specified, applying PowerProfile to all nodes", "profile", profile.Name, "nodeName", nodeName)
 		return true, nil
 	}
 
-	// Get the node to check its labels.
 	node := &corev1.Node{}
-	err := c.Get(context.TODO(), client.ObjectKey{Name: nodeName}, node)
-	if err != nil {
-		// If we can't get the node, don't apply the profile.
+	if err := c.Get(context.TODO(), client.ObjectKey{Name: nodeName}, node); err != nil {
```
Comment (Missxiaoguo) [inline controllers/common.go]: nit: context.TODO is used.

**Review by Missxiaoguo — COMMENTED** [review]
Looks great to me! Just one minor comment. 
/lgtm

File: controllers/powerconfig_controller.go [hunk 2]
```diff
@@ -210,6 +213,11 @@ func (r *PowerConfigReconciler) Reconcile(c context.Context, req ctrl.Request) (
 				return ctrl.Result{}, err
 			}
 		}
+
+		// Write NodeInfo via SSA.
+		if err := r.applyNodeInfo(c, &node, powerNodeStateName, &logger); err != nil {
+			return ctrl.Result{}, err
+		}
```
Comment (bartwensley) [inline controllers/powerconfig_controller.go]: Is this going to update the PowerNodeState for every node every 5s? Should this be only done if the PowerNodeState CR was created? Is there some other way to avoid writing on every reconcile loop?

File: controllers/uncore_controller.go [hunk 3]
```diff
@@ -44,135 +54,331 @@ type UncoreReconciler struct {
...
+		}
+		for _, die := range *pkg.Dies() {
+			if err := die.SetUncore(nil); err != nil {
+				logger.Error(err, "could not reset uncore on die")
+				return err
+			}
+		}
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/uncore_controller.go:287]: - Added nil guards for Topology().Packages() and pkg.Dies() pointer dereferences in resetUncoreSettings, since these return *[]Package / *[]Die and could be nil if the hardware topology is empty.
- For Topology() itself, added a precondition check at the top of Reconcile() alongside the existing namespace and NODE_NAME guards — if the topology isn't available, nothing in the reconcile can work.
- We opted not to add a nil guard for r.PowerLibrary — it's injected at controller setup in main.go, same as r.Client. No controller in this codebase guards against r.Client != nil, and a nil PowerLibrary indicates a broken setup, not a recoverable runtime condition.

File: controllers/uncore_controller.go [hunk 4]
```diff
@@ -44,135 +54,331 @@ type UncoreReconciler struct {
...
+				CreateFunc:  func(e event.CreateEvent) bool { return false },
+				DeleteFunc:  func(e event.DeleteEvent) bool { return false },
+				GenericFunc: func(e event.GenericEvent) bool { return false },
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					if e.ObjectNew.GetName() != nodeName {
+						return false
+					}
+					oldNode := e.ObjectOld.(*corev1.Node)
+					newNode := e.ObjectNew.(*corev1.Node)
+					return !reflect.DeepEqual(oldNode.Labels, newNode.Labels)
```
Comment (bartwensley) [inline controllers/uncore_controller.go]: Comment from Claude (I don't know if it is accurate):

>   reflect.DeepEqual on maps is correct but slower than maps.Equal (Go 1.21+). The label check fires on every node update event. For correctness it's fine, but if this project uses Go 1.21+, maps.Equal would be a cleaner choice.

**Review by bartwensley — COMMENTED** [review]
Looks good - just a couple questions.

File: controllers/powerconfig_controller.go (hunk 2 above)
Comment (irinamihai) [reply controllers/powerconfig_controller.go]: We can write to the PowerNodeState only when it's created. If PowerNodeState is mistakenly modified, the user can delete it and the powerconfig controller will recreate it.

File: controllers/uncore_controller.go (hunk 4 above)
Comment (irinamihai) [reply controllers/uncore_controller.go]: I think it makes more sense from a clarity perspective, since this would be something run only on node label updates. Labels are not that long.
Worth the update though.

File: controllers/uncore_controller.go [hunk 5]
```diff
@@ -44,135 +54,343 @@ type UncoreReconciler struct {
...
 		if err != nil {
-			logger.Error(err, "error setting uncore")
-			return ctrl.Result{Requeue: false}, err
+			applyErrors = append(applyErrors, fmt.Sprintf("error creating system uncore: %v", err))
+		} else if err := r.PowerLibrary.Topology().SetUncore(p_uncore); err != nil {
+			applyErrors = append(applyErrors, fmt.Sprintf("error setting system uncore: %v", err))
+		} else {
+			configParts = append(configParts, fmt.Sprintf("SysMin: %d, SysMax: %d", *spec.SysMin, *spec.SysMax))
 		}
 	}
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/uncore_controller.go:197]: A temporary check would be worth it, until we implement the real solution, which is to use the hardware min/max if either one of SysMin/SysMax is missing.

[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/uncore_controller.go:197]: Yes, open the issue.

**Review by Missxiaoguo — COMMENTED** [review]
/lgtm
/approve
